eessi/testsuite/tests/apps/tensorflow/tensorflow.py

```diff
@@ -0,0 +1,125 @@
+        return sn.assert_eq(n_ranks, self.num_tasks)
+
+    @deferrable
+    def assert_completion(self):
+        '''Assert that the test ran until completion'''
+        n_fit_completed = sn.count(sn.extractall('^Rank [0-9]+: Keras fit completed', self.stdout))
+        
+        return sn.all([
+            sn.assert_eq(n_fit_completed, self.num_tasks),
+        ])
+
+    @deferrable
+    def assert_convergence(self):
+        '''Assert that the network learned _something_ during training'''
+        accuracy=sn.extractsingle('^Final accuracy: (?P<accuracy>\S+)', self.stdout, 'accuracy', float)
+        # mnist is a 10-class classification problem, so if accuracy >> 0.2 the network 'learned' something
+        return sn.assert_gt(accuracy, 0.2)
+
+    @sanity_function
+    def assert_sanity(self):
+        '''Check all sanity criteria'''
+        return sn.all([
+            self.assert_tf_config_ranks(),
+            self.assert_completion(),
+            self.assert_convergence(),
+        ])
+
+    @performance_function('img/s')
+    def perf(self):
+        return sn.extractsingle(r'^Performance:\s+(?P<perf>\S+)', self.stdout, 'perf', float)
+
+    @run_after('init')
+    def run_after_init(self):
+        """hooks to run after the init phase"""
+        hooks.filter_valid_systems_by_device_type(self, required_device_type=self.device_type)
+        hooks.set_modules(self)
+        hooks.set_tag_scale(self)
+
+    @run_after('init')
+    def set_executable_opts(self):
+        """Set executable opts based on device_type parameter"""
+        num_default = 0  # If this test already has executable opts, they must have come from the command line
+        hooks.check_custom_executable_opts(self, num_default=num_default)
+        if not self.has_custom_executable_opts:
+            self.executable_opts += ['--device', self.device_type]
+
+    @run_after('init')
+    def set_test_descr(self):
+        self.descr = f'TensorFlow benchmark on {self.device_type}'
+
+    @run_after('setup')
+    def run_after_setup(self):
+        """hooks to run after the setup phase"""
+        # TODO: implement
+        # It should bind to socket, but different MPIs may have different arguments to do that...
+        # We should at very least prevent that it binds to single core per process, as that results in many threads being scheduled to one core
+        # binding may also differ per launcher used. It'll be hard to support a wide range and still get proper binding
+        if self.device_type == 'cpu':
+            hooks.assign_one_task_per_compute_unit(test=self, compute_unit=DEVICES['CPU_SOCKET'])
+        elif self.device_type == 'gpu':
+            hooks.assign_one_task_per_compute_unit(test=self, compute_unit=DEVICES['GPU'])
+        else:
+            raise NotImplementedError(f'Failed to set number of tasks and cpus per task for device {self.device_type}')
+
+    @run_after('setup')
+    def set_binding_policy(self):
```

Comment:
We also enforce this. We have 2 socket nodes, and users can allocate 1/8th of a node as smallest entity. In SLURM, we configure nodes to have 8 sockets, and then tell SLURM to hand out allocations per socket. The number of sockets (8) doesn't reflect reality, but it does force SLURM to allocate consecutive sets of cores.

Of course, we cannot assume _anything_ about how this is done on different systems. I actually found out I already made a hook for this process binding (`set_compact_process_binding`), I have no idea why I wasn't using it. I showed it in the meeting as something that would work also for hyperthreading (which I think it does, but will have to check), but I think it should do a sensible binding for _any_ task, whether it is hybrid, or pure MPI. I'll use it for now, and do some tests to see if that gives me the correct binding behaviour.

In terms of how this will work on systems that _don't_ hand out consecutive sets of cores: I don't know, that depends on how mpirun and srun handle that. The only way to know for sure is to try it on a system that has this. But I think this is a case of: let's just deploy it, and we'll get an issue at some point if this occurs. Then we fix it :)